prov/rxm/src/rxm_cq.c

```diff
@@ -409,25 +409,24 @@ ssize_t rxm_cq_handle_data(struct rxm_rx_buf *rx_buf)
 
 static ssize_t rxm_handle_recv_comp(struct rxm_rx_buf *rx_buf)
 {
-	struct rxm_recv_match_attr match_attr;
+	struct rxm_recv_match_attr match_attr = {
+		.addr = FI_ADDR_UNSPEC,
+	};
 	struct dlist_entry *entry;
-	struct util_cq *util_cq;
-
-	util_cq = rx_buf->ep->util_ep.rx_cq;
 
-	if ((rx_buf->ep->rxm_info->caps & (FI_SOURCE | FI_DIRECTED_RECV)) &&
-	    !rx_buf->conn) {
-		rx_buf->conn = rxm_key2conn(rx_buf->ep, rx_buf->pkt.ctrl_hdr.conn_id);
-		if (OFI_UNLIKELY(!rx_buf->conn))
-			return -FI_EOTHER;
-		match_attr.addr = rx_buf->conn->handle.fi_addr;
-	} else {
-		match_attr.addr = FI_ADDR_UNSPEC;
+	if (rx_buf->ep->rxm_info->caps & FI_SOURCE) {
+		if ((rx_buf->ep->rxm_info->caps & FI_DIRECTED_RECV) && !rx_buf->conn) {
+			rx_buf->conn = rxm_key2conn(rx_buf->ep,
+						    rx_buf->pkt.ctrl_hdr.conn_id);
+			if (OFI_UNLIKELY(!rx_buf->conn))
+				return -FI_EOTHER;
+			match_attr.addr = rx_buf->conn->handle.fi_addr;
+		}
+		rx_buf->ep->util_ep.rx_cq->src[
+			ofi_cirque_windex(rx_buf->ep->util_ep.rx_cq->cirq)] =
+				rx_buf->conn->handle.fi_addr;
```

Comment:
This still doesn't look right. rx_buf->conn is set only when FI_DIRECTED_RECV is set. So the above code could result in a NULL pointer dereference.

Reply:
It seems to me that this doesn't change the previous behavior.
You're right, this seems a bug. I'll evaluate this.